optuna/storages/_rdb/models.py

```diff
@@ -447,23 +449,82 @@ def where_trial_id(cls, trial_id: int, session: orm.Session) -> List["TrialValue
 
     __tablename__ = "trial_intermediate_values"
     __table_args__: Any = (UniqueConstraint("trial_id", "step"),)
     trial_intermediate_value_id = Column(Integer, primary_key=True)
     trial_id = Column(Integer, ForeignKey("trials.trial_id"), nullable=False)
     step = Column(Integer, nullable=False)
     intermediate_value = Column(Float(precision=FLOAT_PRECISION), nullable=True)
-    intermediate_value_type = Column(Enum(FloatTypeEnum), nullable=False)
+    intermediate_value_type = Column(Enum(TrialIntermediateValueType), nullable=False)
 
     trial = orm.relationship(
         TrialModel, backref=orm.backref("intermediate_values", cascade="all, delete-orphan")
     )
 
+    @classmethod
+    def _intermediate_value_to_stored_repr(
+        cls,
+        value: float,
+    ) -> Tuple[Optional[float], TrialIntermediateValueType]:
+        if math.isnan(value):
+            return (None, cls.TrialIntermediateValueType.NAN)
+        elif value == float("inf"):
+            return (None, cls.TrialIntermediateValueType.INF_POS)
+        elif value == float("-inf"):
+            return (None, cls.TrialIntermediateValueType.INF_NEG)
+        else:
+            return (value, cls.TrialIntermediateValueType.FINITE)
+
+    @classmethod
+    def _stored_repr_to_intermediate_value(
+        cls, value: Optional[float], float_type: TrialIntermediateValueType
+    ) -> float:
+        if float_type == cls.TrialIntermediateValueType.NAN:
+            assert value is None
+            return float("nan")
+        elif float_type == cls.TrialIntermediateValueType.INF_POS:
+            assert value is None
+            return float("inf")
+        elif float_type == cls.TrialIntermediateValueType.INF_NEG:
+            assert value is None
+            return float("-inf")
+        else:
+            assert float_type == cls.TrialIntermediateValueType.FINITE
+            assert value is not None
+            return value
+
+    def __init__(self, *, original_intermediate_value: float, **kwargs: Any) -> None:
+        (
+            sanitized_intermediate_value,
+            intermediate_value_type,
+        ) = self._intermediate_value_to_stored_repr(original_intermediate_value)
+        super().__init__(
+            **kwargs,
+            intermediate_value=sanitized_intermediate_value,
+            intermediate_value_type=intermediate_value_type,
+        )
+
+    @property
+    def original_intermediate_value(self) -> float:
+        return self._stored_repr_to_intermediate_value(
+            self.intermediate_value, self.intermediate_value_type
+        )
+
+    @original_intermediate_value.setter
+    def original_intermediate_value(self, value: float) -> None:
+        (
+            sanitized_intermediate_value,
+            intermediate_value_type,
+        ) = self._intermediate_value_to_stored_repr(value)
+        self.intermediate_value = sanitized_intermediate_value
+        self.intermediate_value_type = intermediate_value_type
```

Comment:
Would hiding `self.intermediate_value` etc. as `self._intermediate_value` be okay?
This is possible, but I wasn't sure if that was worth the magic.

Reply:
What do you think? @toshihikoyanase

Reply:
> Would hiding `self.intermediate_value` etc. as `self._intermediate_value` be okay?

I don't think so. I guess keeping the one-to-one relationship between columns and member variables will reduce the complexity of code while it increase the code complexity of representation conversion.

Reply:
But `intermediate_value` and `intermediate_value_type` are logically one column, they just cannot be one column due to limitations in DB side. I think it is reasonable to hide this complexity in the `Model` class.

Reply:
I changed the properties to functions. It would be more clear that the code is not directly writing to a column in DB but is actually doing a conversion. PTAL. @toshihikoyanase

Reply:
Let me merge this conversation with @HideakiImamura -san's.